blog/_posts/2022-04-04-a-dev-environment-repl.md

```diff
@@ -0,0 +1,132 @@
+---
+slug: "a-dev-environment-repl"
+date: 2022-04-04
+author: siegs
+layout: blog
+title: A REPL for your dev environment
+image: /assets/images/a-dev-environment-repl/curly.jpg
+image_caption: '"Curly" by <a href="https://www.flickr.com/photos/gammaman/6212822745/">Eli Christman</a>'
+description: How Tilt can power experimentation and a fast feedback loop
+tags:
+- repl
+- development
+- dev environments
+- devtools
+---
+Let's spend a moment pondering the foundation of software tools: the humble, trusty REPL. Call it by whatever name you like: console, terminal, shell, sandbox, debugger, scratch pad. The concept is almost as old as computing itself: dating back to an [early LISP implementation on the PDP-1 computer][lisp-pdp1], a Read-Eval-Print-Loop program is the essence of interactive computing.
+
+Many of us started programming with a REPL, whether we knew it at the time or not. (My own introduction came via Logo and Basic on the Apple II.) The thrill of entering obscure abbreviations and commands into the computer and receiving instant feedback can be a big dopamine hit. It can feel like playing a game with infinite lives, where you always have another chance to slay the boss 🐉.
+
+REPLs encourage experimentation and learning by efficiently providing a tight feedback loop (almost literally; the PL in REPL being "Print Loop", where "print" serves as the feedback mechanism). Good REPLs help with exploration by providing completion and context awareness. Here's a simple example in Ruby:
+
+![Ruby IRB Fibonacci example](/assets/images/a-dev-environment-repl/irb-fib.gif)
+
+Some languages use REPLs as the focal point for writing code (see Smalltalk, and to an extent, any member of the Lisp family), while others (C/C++ and many compiled languages) don't come with any built-in REPL at all. More recently, online services let you try a language or share runnable snippets without having to download any software at all (see [Ruby][tryruby], [Go][goplay], and [JavaScript][repljs]).
+
+Tilt, as a tool for assembling development environments, is all about tight [feedback loops][controlloop]. Tilt responds to changes in your environment in the same way that the REPL snaps to action when you hit ENTER. One of the events Tilt responds to is changes to the `Tiltfile`, Tilt's own [Starlark][starlark]-based configuration file. With a little creativity and flexibility, we can run Tilt in a terminal inside or next to our editor (`tilt up --stream`) and mimic the feedback and responsiveness of a REPL:
+
+![Tilt Fibonacci example](/assets/images/a-dev-environment-repl/tilt-fib.gif)
+
+The result is a loop that feels surprisingly responsive: write a bit of code, hit `Ctrl-S` or `⌘-S`, and see the result! If you're exploring Tilt and haven't used Starlark (or Python, Starlark's parent language), this can be a great way to fool around and get your bearings with tiltfile syntax or builtins.
```

Comment:
```suggestion
The result is a loop that feels surprisingly responsive: write a bit of code, hit `Ctrl-S` or `⌘-S`, and see the result! If you're exploring Tilt and haven't used Starlark (or Python, Starlark's parent language), this can be a great way to fool around and get your bearings with `Tiltfile` syntax or builtins.
```